numpy/_core/numerictypes.py

```diff
@@ -359,6 +360,87 @@ def issubsctype(arg1, arg2):
     return issubclass(obj2sctype(arg1), obj2sctype(arg2))
 
 
+@set_module('numpy')
+def isdtype(dtype, kind):
+    """
+    Returns a boolean indicating whether a provided dtype is of a specified
+    data type ``kind``.
+
+    Parameters
+    ----------
+    dtype : dtype
+        The input dtype.
+    kind : dtype or str or tuple of dtypes/strs.
+        dtype or dtype kind. Allowed dtype kinds are:
+        * ``bool`` : boolean kind
+        * ``signed integer`` : signed integer data types
+        * ``unsigned integer`` : unsigned integer data types
+        * ``integral`` : integer data types
+        * ``real floating`` : real-valued floating-point data types
+        * ``complex floating`` : complex floating-point data types
+        * ``numeric`` : numeric data types
+
+    Returns
+    -------
+    out : bool
+
+    See Also
+    --------
+    issubdtype
+
+    Examples
+    --------
+    >>> import numpy as np
+    >>> np.isdtype(np.int64, (np.uint64, "signed integer"))
+    True
+    >>> np.isdtype(np.float32, np.float64)
+    False
+
+    """
+    # validate and preprocess arguments
+    if not isinstance(dtype, (type, ma.dtype)):
```

Comment:
I'm not sure if there's an easy way to spell this that's more accurate than just checking for `type`, but this would let you do e.g. `np.isdtype(np.int64, dict)` and return False. I guess that is true but is perhaps against the spirit of this check. I don't know if there's an easy way to get a list of all the dtype scalar types numpy knows about at any given moment (which could include user dtypes or dtypes defined in downstream packages).

Reply:
I don't know it either. Let's ask about it during today's community call.

Reply:
Maybe checking `issubclass(dtype, np.generic)` would make snese?

```python
In [31]: np.int64.mro()
Out[31]:
[numpy.int64,
 numpy.signedinteger,
 numpy.integer,
 numpy.number,
 numpy.generic,
 object]
```

Reply:
> reject all dtypes or scalar types that aren't in the array API standard

That would make it harder to use while supporting, e.g., `float16`, and seems like an odd suggestion - what is the justification?

Reply:
We couldn’t up with another way to satisfy the API contract and support dtypes defined outside numpy. If we could know whether an arbitrary python type is associated with a dtype defined outside numpy, we’d get partway there, but then we’d need to be able to classify it using the categories this function supports.

At the meeting @mattip suggested that since this function is specifically for array API support and all of these questions about dtypes are well-defined using the array API dtypes, that it may be simplest to reject others.

Do you have an alternate approach? Ideally numpy wouldn’t be returning incorrect answers here for a dtype defined outside numpy, I think.

Reply:
The design is perfectly well specified for all numpy dtypes, so returning the expected result for them is desired. This API is supposed to make sense for NumPy itself too, not only for code specifically written for cross-library array API standard libs. 

For third-party dtypes there is still a lot in flux and the new APIs for creating them haven't even been exposed yet, so that is a completely separate question to me. Such dtypes are also quite rare, so it isn't nearly as important.

That said, `isdtype` should be able to support the third-party dtypes that `np.issubdtype` supports as well I'd think. So that would be the thing to discuss, is that complete and desirable. Options:
- use the same checks as `issubdtype` and include the results (e.g, isdtype(dt, 'unsigned integer')` could use check using the `np.unsignedinteger` ABC)
- document that all third-party dtypes are not handled, and either return `False` or raise for those.
    - either way, for builtin dtypes, `isdtype` should not raise exceptions

Reply:
Looking at the `jax` `bfloat16` type as a real-world example, I see:

```
In [10]: np.issubdtype(ml_dtypes.bfloat16, np.floating)
Out[10]: False

In [11]: ml_dtypes.bfloat16.mro()
Out[11]: [ml_dtypes.bfloat16, numpy.generic, object]
```

Which isn't particularly useful but is correct according to the implementation of `issubdtype`.

So I think we might need to just not handle third-party dtypes for now, until we have a better story for registering dtypes in a type hierarchy. Does that sound reasonable?

Handling all the builtin numpy dtypes here makes sense, no worries about doing that.

Reply:
> So I think we might need to just not handle third-party dtypes for now, until we have a better story for registering dtypes in a type hierarchy. Does that sound reasonable?

Thanks for checking. That sounds fine to me.

Reply:
For what it's worth, the reason that `np.issubdtype(ml_dtypes.bfloat16, np.floating)` is that in several places numpy hard-codes the assumption that there is only a single 16-bit floating-point type, so we could not make `bfloat16` a subclass of `np.floating` without causing some collisions across the codebase. If you have suggestions for how to do better, we'd love to hear it!

Reply:
I’d love to relax that assumption inside numpy, I’m sure patches that do that would be reviewed if you can get anyone to work on it. I also wasn’t privy to previous discussions about this but at this point I think numpy should probably support bfloat16 natively. Of course that’s just my opinion. Also with the new dtype API it should be much more straightforward to upstream a new dtype into numpy, there’s no need anymore to mess with all the complicated custom templating and codegen.